fem/coefficient.hpp

```diff
@@ -1260,6 +1302,15 @@ class SymmetricMatrixCoefficient : public MatrixCoefficient
    /// Get the size of the matrix.
    int GetSize() const { return height; }
 
+   /// @brief Fill the QuadratureFunction @a qf by evaluating the coefficient at
+   /// the quadrature points.
+   ///
+   /// @note As opposed to MatrixCoefficient::Eval, this function stores only
+   /// the @a symmetric part of the matrix at each quadrature point.
+   ///
+   /// The @a vdim of the coefficient should be equal to height*(height+1)/2.
+   virtual void ProjectSymmetric(QuadratureFunction &qf);
```

Comment:
The reason is because `SymmetricMatrixCoefficient` is derived from `MatrixCoefficient`. The kernel may handle the special case of symmetric coefficients stored in symmetric layout (with ~half the memory), in which case you want `ProjectSymmetric`. However, if the kernel doesn't handle this special case, and just handles a general matrix coefficient, you would want to use the full matrix storage — less implementation effort, but more memory use/access.

This is chosen with `CoefficientStorage`. If you have `CoefficientStorage::SYMMETRIC`, the `CoefficientVector` will use `ProjectSymmetric `, otherwise it will use `Project`.

This is related to what's written in the Doxygen comment, but there was a typo that made it confusing.